integration-tests/protocolv2_test.go

```diff
@@ -0,0 +1,171 @@
+package integration_tests
+
+import (
+	"context"
+	"fmt"
+	"github.com/datastax/go-cassandra-native-protocol/client"
+	"github.com/datastax/go-cassandra-native-protocol/datatype"
+	"github.com/datastax/go-cassandra-native-protocol/frame"
+	"github.com/datastax/go-cassandra-native-protocol/message"
+	"github.com/datastax/go-cassandra-native-protocol/primitive"
+	"github.com/datastax/zdm-proxy/integration-tests/setup"
+	"github.com/stretchr/testify/require"
+	"net"
+	"testing"
+)
+
+func TestProtocolV2Connect(t *testing.T) {
+	originAddress := "127.0.0.2"
+	targetAddress := "127.0.0.3"
+
+	serverConf := setup.NewTestConfig(originAddress, targetAddress)
+	proxyConf := setup.NewTestConfig(originAddress, targetAddress)
+	proxyConf.ControlConnMaxProtocolVersion = "3" // simulate protocol downgrade to V2
+
+	testSetup, err := setup.NewCqlServerTestSetup(t, serverConf, false, false, false)
+	require.Nil(t, err)
+	defer testSetup.Cleanup()
+
+	originRequestHandler := NewProtocolV2RequestHandler("origin", "dc1", "127.0.0.4")
+	targetRequestHandler := NewProtocolV2RequestHandler("target", "dc1", "127.0.0.5")
+
+	testSetup.Origin.CqlServer.RequestHandlers = []client.RequestHandler{
+		originRequestHandler.HandleRequest,
+		client.NewDriverConnectionInitializationHandler("origin", "dc1", func(_ string) {}),
+	}
+	testSetup.Target.CqlServer.RequestHandlers = []client.RequestHandler{
+		targetRequestHandler.HandleRequest,
+		client.NewDriverConnectionInitializationHandler("target", "dc1", func(_ string) {}),
+	}
+
+	err = testSetup.Start(nil, false, primitive.ProtocolVersion2)
+	require.Nil(t, err)
+
+	proxy, err := setup.NewProxyInstanceWithConfig(proxyConf) // starts the proxy
+	if proxy != nil {
+		defer proxy.Shutdown()
+	}
+	require.Nil(t, err)
+}
+
+func TestProtocolV2Query(t *testing.T) {
```

Comment:
I'd say we can probably merge these 2 tests into a single one that connects and sends a query